framework/sandstone.cpp

```diff
@@ -1045,18 +1045,24 @@ namespace {
 class QueueDeviceSchedule : public DeviceSchedule
 {
 public:
-    int get_next_cpu() override
-    {
-        // Returns next CPU from the queue
+    void device_reschedule() override {
+        // Select a cpu from the queue
         std::lock_guard lock(q_mutex);
         if (q_idx == 0)
             shuffle_queue();
-        int result = queue[q_idx];
+
+        int next_cpu = queue[q_idx];
         if (++q_idx == queue.size())
             q_idx = 0;
-        return result;
+
+        if (!pin_to_logical_processor(LogicalProcessor(cpu_info[next_cpu].cpu_number))) {
+            log_debug("Failed to reschedule %d (%ld) to CPU %d", thread_num, pthread_self(), next_cpu);
```

Comment:
(Pre-existing but please fix)
This should be a `log_warn`.

The formatter for `pthread_self()` should be `%tu` and you should cast it to `uintptr_t` so it will compile everywhere.
Linux: `typedef unsigned long int pthread_t;`
FreeBSD: `typedef struct  pthread                 *pthread_t;`
Winpthreads: `typedef uintptr_t pthread_t;`

You may want to put the actual pinning and warning in the base class, in a protected function, to avoid duplication.

Reply:
I was thinking that probably would be better to simply print the `tid_t tid` that each thread stores. In terms of debugging, I don't know which would be more useful